### PR TITLE
Use Guava Cache when creating `LimitedMap` rather than `LinkedHashMap`

### LINKED ISSUE #2612: State Cache Size Exceeding Limit
### Description
After running overnight a number of nodes are struggling with excessive GC and reporting that there are thousands of states in memory.  The limit should be 160 cached states.

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Switched `LimitedMap` to create a Guava `Cache` instead of a `LinkedHashMap`.
Concurrent access safe by default (and much more efficiently than using synchronizedMap) and will fix an issue where LinkedHashMap was unexpectedly exceeding it's expected maximum size.
Only least recently used eviction is supported. This was used in all but one place which never accessed individual items anyway so is unaffected by the change.

## Fixed Issue(s)
fixes #2612 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java [hunk 1]
```diff
@@ -47,10 +45,12 @@ public OperationPool(Class<T> clazz, OperationStateTransitionValidator<T> operat
 
   public SSZList<T> getItemsForBlock(BeaconState stateAtBlockSlot) {
     SSZMutableList<T> itemsToPutInBlock =
-        SSZList.createMutable(clazz, maxNumberOfElementsInBlock.get(clazz));
+        SSZList.createMutable(clazz, MAX_NUMBER_OF_ELEMENTS_IN_BLOCK.get(clazz));
+    // Note that iterating through all items does not affect their access time so we are effectively
+    // evicting the oldest entries when the size is exceeded as we only ever access via iteration.
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java:50]: is it worth investigating if we can explicitly make a call to update the accessed time of what we actually want to use?

**Review by rolfyone — APPROVED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java:50]: We used to deliberately use evict oldest policy here so this is actually what we wanted.

**Comment by Nashatyrev** [comment]
@ajsutton, I'd like to return back the `LinkedHashMap` as it's ~3x-4x more efficient than Guava cache after fixing the weak `Pair.hashCode()`. Do you have any clue on what was exactly wrong with the `LinkedHashMap`?

**Comment by ajsutton** [comment]
> @ajsutton, I'd like to return back the `LinkedHashMap` as it's ~3x-4x more efficient than Guava cache after fixing the weak `Pair.hashCode()`. Do you have any clue on what was exactly wrong with the `LinkedHashMap`?

The issue was with thread safety - we were concurrently modifying the map which is a) completely unsafe and b) resulted in `LinkedHashMap` growing unbounded because it assumes only one item can be added at a time and doesn't handle multiple threads adding at the same time.  So we can switch back from Cache but then we'd have to use something like `Collections.synchronizedMap` which I suspect would be even slower.

**Comment by Nashatyrev** [comment]
That's weird cause prior to this PR we were effectively using `Collections.synchronizedMap(new LinkedHashMap())`
I think I could try testing concurrency on pre-PR revision and see if I could reproduce the problem... 

> `Collections.synchronizedMap` which I suspect would be even slower.

I was testing with `synchronizedMap` and it shown ~x3-x4 performance gain comparing to Guava `Cache` on all operations: `put`,`get` and `copy`.  Funny fact that with the weak `Pair.hashCode()` Guava `Cache` indeed works faster

**Comment by ajsutton** [comment]
> That's weird cause prior to this PR we were effectively using `Collections.synchronizedMap(new LinkedHashMap())`

Previously we had both `LimitedMap` which returned a plain `LinkedHashMap` and wasn't thread safe and `ConcurrentLimitedMap` which used `synchronizedMap`.  If we'd always used `ConcurrentLimitedMap` it would have been fine but in some key places we got it wrong.

If `synchronizedMap(new LinkedHashMap())` is faster for us than `Cache` then there's no problem switching to that.

**Comment by Nashatyrev** [comment]
(just for the record) Switching back to `ConcurrentLimitedMap` here #3758

**Pull request merged**